### DISCUSSION TITLE
"No text was generated from this prompt" on Quest 2 VR device

### DISCUSSION THREAD

**Discussion by CadencePr** [srcnalt/OpenAI-Unity#21] [Q&A] [answered]
Thank you for sharing a fantastic project!
I was able to make it run well on the Unity editor. However, when it runs on Meta Quest 2 device (Android), I always get "No text was generated from this prompt." Double-checked API Keys and internet access permission but found no issues. (When I intentionally put the wrong API key, it gives me an API key error. Also, other components in my app require internet access and work well) 
Any suggestions would be appreciated!

**Comment by srcnalt** [comment]
Hi @CadencePr, have you checked if the same prompt gave you the result in the editor? The message is from OpenAI API itself it seems, my only guess is it got stuck in OpenAI filters. Even empty prompts get replies usually.

**Reply by CadencePr** [reply]
Yes, it works very well in the Unity editor's play mode. (including play mode running on the headset connected with USB-C Link) Getting no response only when I build as .apk and deployed it to the device. Since it gives me auth error when I put the wrong API keys, it seems the internet connection/permission is fine. The message is from ChatGPT.cs file which means there is no completionResponse. Curious about what could make any difference on the Quest device vs the editor other than internet access permission. 

            if (completionResponse.Choices != null && completionResponse.Choices.Count > 0)
            {
                textArea.text = completionResponse.Choices[0].Text;
                Instruction += $"{completionResponse.Choices[0].Text}\nQ: ";
            }
            else
            {
                Debug.LogWarning("No text was generated from this prompt.");
            }

**Reply by srcnalt** [reply]
I don't think meta would block apps making web requests 🤔

**Reply by divinitize** [reply]
did you find a solution to this? i have the same issue on android

**Reply by srcnalt** [reply]
Hi @divinitize just saw your youtube comment. Which version of the package are you using, I suspect this might be a system language related issue, which was discussed here #14

**Reply by divinitize** [reply]
0.1.10 is the one im using

**Reply by srcnalt** [reply]
How about mobile device system language, is it English? If not, please change it to English and try. I am almost sure it could not match field names and caused the issues. You might al well add a debug line into DispatchRequest methods in OpenAiApi.cs to print out the downloadHandler text before it's being parsed, so we can see if you do receive any response.

**Reply by divinitize** [reply]
yes it's english UK, im not sure how to add a debug log to that as it's not a monobehaviour im guessing?

**Reply by srcnalt** [reply]
You can find the package in Packages folder in Project window, you can see the entire source code there.

**Reply by divinitize** [reply]
Yes i understand, i have edited the OpenAIApi already for other things, i mean im not sure how to log the dispatchrequest like you mentioned

**Reply by srcnalt** [reply]
Can you add `Debug.Log(request.downloadHandler.text)` before this line?

https://github.com/srcnalt/OpenAI-Unity/blob/d347eeaeb5dbae0d2c36ba89f083927a77c34845/Runtime/OpenAIApi.cs#LL79C57-L79C85

**Reply by divinitize** [reply]
done, doing a new build for android, standby

**Reply by divinitize** [reply]
ok i think this is the right debug log in android logcat because i dont recall seeing it before but im getting 

`{
 "error": {
 "message": "you must provide a model parameter",
 "type": "invalid_request_error",
"param": null,
 "code": null
}
 }
`

**Reply by divinitize** [reply]
ill also just add im using 

`// Complete the instruction
            var completionResponse = await openai.CreateChatCompletion(new CreateChatCompletionRequest()
            {
                Model = "gpt-3.5-turbo-0301",
                Messages = messages
            });`

**Reply by srcnalt** [reply]
Your request looks okay, but it acts as if you didn't set the model field. Can you also print out the json var in here right under this line, do we can see if the payload is broken.

https://github.com/srcnalt/OpenAI-Unity/blob/d347eeaeb5dbae0d2c36ba89f083927a77c34845/Runtime/OpenAIApi.cs#L134

**Reply by divinitize** [reply]
sorry do i just need Debug.Log(json); or anything else like .text?

**Reply by divinitize** [reply]
i added a some debug logs before and after so i could see where to copy and paste and got this 

`2023/03/31 11:47:13.130 9494 9524 Info Unity OpenAI.OpenAIApi:CreatePayload(T)
2023/03/31 11:47:13.130 9494 9524 Info Unity OpenAI.<CreateChatCompletion>d__13:MoveNext()
2023/03/31 11:47:13.130 9494 9524 Info Unity System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Start(TStateMachine&)
2023/03/31 11:47:13.130 9494 9524 Info Unity OpenAI.OpenAIApi:CreateChatCompletion(CreateChatCompletionRequest)
2023/03/31 11:47:13.130 9494 9524 Info Unity OpenAI.<SendReply>d__21:MoveNext()
2023/03/31 11:47:13.130 9494 9524 Info Unity System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start(TStateMachine&)
2023/03/31 11:47:13.130 9494 9524 Info Unity OpenAI.ChatGPT:SendReply()
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.Events.UnityAction:Invoke()
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.Events.UnityEvent:Invoke()
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.EventSystems.EventFunction`1:Invoke(T1, BaseEventData)
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.EventSystems.ExecuteEvents:Execute(GameObject, BaseEventData, EventFunction`1)
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.EventSystems.StandaloneInputModule:ProcessTouchPress(PointerEventData, Boolean, Boolean)
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.EventSystems.StandaloneInputModule:ProcessTouchEvents()
2023/03/31 11:47:13.130 9494 9524 Info Unity UnityEngine.EventSystems.StandaloneInputModule:Process()
`

EDIT: infact i dont think it is sending anything, as from this 

Debug.Log("LOOK HERE" + json);
Logcat is sending 
2023/03/31 11:47:13.130 9494 9524 Info Unity LOOK HERE{}
which looks like its just empty opening and closing curcly brackets

but when i do it in the editor it actually shows information
LOOK HERE{"model":"gpt-3.5-turbo-0301","messages":[{"role":"assistant","content":"My response is here"},{"role":"user","content":"test"}],"temperature":1.0,"n":1,"stream":false,"presence_penalty":0.0,"frequency_penalty":0.0}

**Reply by srcnalt** [reply]
Could you check the version of NewtonsoftJson you are using from Window > Package Manager menu? 
This is a platform related json parsing issue.

**Reply by divinitize** [reply]
3.0.2

**Reply by srcnalt** [reply]
JsonConvert.SerializeObject(request, jsonSerializerSettings); returns an empty object for some reason.

**Reply by srcnalt** [reply]
Okay that is the correct version, could you check if request is not null in android, you could debug.log it right before json parsing line, maybe output the model, like `Debug.Log(request.model);`

**Reply by divinitize** [reply]
Incase this is any relevant information too

in my android player settings

min api level is set to android 6.0
target api is automatic (highest installed)
scripting backend is IL2CPP
Api compatibility level is .NET 4.x

**Reply by divinitize** [reply]
editor error

'T' does not contain a definition for 'model' and no accessible extension method 'model' accepting a first argument of type 'T' could be found (are you missing a using directive or an assembly reference?)

**Reply by divinitize** [reply]
logging just "request" only gives 

OpenAI.CreateChatCompletionRequest

**Reply by srcnalt** [reply]
okay so its not null, sorry I missed that you'd need to cast it to chat completion object first. But it seems like it is Android + Json parsing issue. Tonight I can try to replicate it myself on my phone, however it did work on mobile for me I can try using these settings. Meanwhile, could simply build the sample and share your .apk file with me so I could also take a look?

**Reply by divinitize** [reply]
yes do you have discord? Dreckly#3291

**Reply by divinitize** [reply]
or let me know how else i can send the apk

**Reply by divinitize** [reply]
did you manage to try this out at all? cheers.

**Reply by srcnalt** [reply]
You can share it here
https://discord.gg/5UhSs8Fq

**Reply by divinitize** [reply]
Sent you a message thanks, will update here if i get a solution to my problem for future viewers.

**Reply by divinitize** [reply]
SOLVED - atleast for now, the issue is using IL2CPP as the scripting backend, changing it to mono fixed the issue. I am unsure why for sure but @srcnalt found this thread https://github.com/jilleJr/Newtonsoft.Json-for-Unity/issues/132 which is likely the cause.
Many thanks again

**Reply by srcnalt** [reply]
In the latest version NewtonsoftJson version is updated and the issues should be solved now.

**Reply by divinitize** [reply]
I updated to 0.1.12 , changed to IL2CPP and still have the same issue "no text was generated by this prompt"
Do i also need to manually update NewtonsoftJson or anything or is it just included in the git?

**Reply by srcnalt** [reply]
Hi @divinitize that would be great if you could manually update, as a dependency it should be updated too, but better be sure. I will also try building for Quest and see if I can replicate this.

**Reply by divinitize** [reply]
I dont seem to have an update option, my Newtonsoft is showing 3.1.0 (synced to 13.0.2).
And i don't know if the build to quest was for this general thread or not, but my issues are on just android mobile. Cheers

**Reply by divinitize** [reply]
My issue stopping me from using ChatGPT feature using IL2CPP on android appears to have been resolved by upgrading my project to the latest unity version, in my case 2021.3.15f1. Only done initial tests so far but it appears to be working. will update if it goes wrong again.

Remember to backup your project if you're doing the same.

**Reply by srcnalt** [reply]
Just built it in Oculus Quest 2 using XR Interaction Toolkit package's Hasns Integration Demo + OpenAI Unity's Streaming Responses demo combined, and I get the response without any issue. Unity 2021.3.5f1

![52251948-87b9-43b1-87aa-d50f6d421eb6](https://user-images.githubusercontent.com/3163281/234124308-d1cc37bd-20f3-45f5-a0b8-0cd0dc31136b.jpg)

**Reply by monurgungor** [reply]
Hey I don't want to bother you on the same topic again but I am really stuck here. I have the same problem as the project works well on the editor but I get the error "Exception: No completion choices returned from the ChatGPT" playing the Quest2 build. Tried every solution above but it's not working.
![image](https://github.com/srcnalt/OpenAI-Unity/assets/69249527/634b8eb1-21e5-4767-a558-5d17a660862b)
It shows the package version as 0.1.12 but I manually updated everything to 0.1.13, I also tried the latest version of Newtonsoft Json and changing scripting backend to both IL2CPP and Mono but no luck at all. Am I missing something here? Unity version is 2021.3.27f1
![image](https://github.com/srcnalt/OpenAI-Unity/assets/69249527/38e83118-11ef-4d05-84d9-348846e6b5d0)

**Comment by srcnalt** [comment, marked as answer]
Just released #68 v0.1.14 which helps with code stripping issues. Could you check and let me know if it works fine?

**Reply by monurgungor** [reply]
Yes, after updating the package from Package Manager and using IL2CPP, everything works great now. Thank you so much.

**Reply by srcnalt** [reply]
Glad to hear it!

**Reply by xmaymo** [reply]
Hi guys, I have a similar problem, in the editor everything works fine but on the Quest build it doesn't respond. 
var response = await openAI.CreateChatCompletion(request) it is not called. 
The openAI credentials are on the code, so I guess there's another problem.

Is there something I need to change to make it work for VR? Any help would be appreciate it!

**Reply by alecpla1** [reply]
@srcnalt have you tested recently? I'm using the below format and it works on PC, but not on Android with the same exact credentials. Previously it was working on Android, stopped maybe weeks/months ago, not sure. 

 openAI = new OpenAIApi(apiKey, organizationID);
